pkg/storage/stores/indexshipper/uploads/table.go

```diff
@@ -0,0 +1,137 @@
+	Upload(ctx context.Context) error
+	Cleanup(indexRetainPeriod time.Duration) error
+	Stop()
+}
+
+// table is a collection of multiple dbs created for a same table by the ingester.
+// All the public methods are concurrency safe and take care of mutexes to avoid any data race.
+type table struct {
+	name                                 string
+	uploader                             string
+	baseUserIndexSet, baseCommonIndexSet storage.IndexSet
+	logger                               log.Logger
+
+	indexSet    map[string]IndexSet
+	indexSetMtx sync.RWMutex
+}
+
+// NewTable create a new table instance.
+func NewTable(name, uploader string, storageClient storage.Client) Table {
+	return &table{
+		name:               name,
+		uploader:           uploader,
+		baseUserIndexSet:   storage.NewIndexSet(storageClient, true),
+		baseCommonIndexSet: storage.NewIndexSet(storageClient, false),
+		logger:             log.With(util_log.Logger, "table-name", name),
+		indexSet:           make(map[string]IndexSet),
+	}
+}
+
+// Name returns the name of the table.
+func (lt *table) Name() string {
+	return lt.name
+}
+
+// AddIndex adds a new index to the table.
+func (lt *table) AddIndex(userID string, idx index.Index) error {
+	lt.indexSetMtx.Lock()
+	defer lt.indexSetMtx.Unlock()
+
+	if _, ok := lt.indexSet[userID]; !ok {
+		baseIndexSet := lt.baseUserIndexSet
+		if userID == "" {
+			baseIndexSet = lt.baseCommonIndexSet
+		}
+		idxSet, err := NewIndexSet(lt.name, userID, baseIndexSet, loggerWithUserID(lt.logger, userID))
+		if err != nil {
+			return err
+		}
+
+		lt.indexSet[userID] = idxSet
+	}
+
+	lt.indexSet[userID].Add(idx)
+	return nil
+}
+
+// ForEach iterates over all the indexes belonging to the user.
+func (lt *table) ForEach(userID string, callback index.ForEachIndexCallback) error {
+	lt.indexSetMtx.RLock()
+	defer lt.indexSetMtx.RUnlock()
+
+	idxSet, ok := lt.indexSet[userID]
+	if !ok {
+		return nil
+	}
+
+	return idxSet.ForEach(callback)
+}
+
+// Upload uploads the index to object storage.
+func (lt *table) Upload(ctx context.Context) error {
+	lt.indexSetMtx.RLock()
+	defer lt.indexSetMtx.RUnlock()
+
+	for _, indexSet := range lt.indexSet {
+		if err := indexSet.Upload(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Cleanup removes indexes which are already uploaded and have been retained for period longer than indexRetainPeriod since they were uploaded.
+func (lt *table) Cleanup(indexRetainPeriod time.Duration) error {
+	lt.indexSetMtx.RLock()
+	defer lt.indexSetMtx.RUnlock()
+
+	for _, indexSet := range lt.indexSet {
+		if err := indexSet.Cleanup(indexRetainPeriod); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Stop closes all the open dbs.
+func (lt *table) Stop() {
+	lt.indexSetMtx.Lock()
+	defer lt.indexSetMtx.Unlock()
+
+	for _, indexSet := range lt.indexSet {
+		indexSet.Close()
+	}
+}
+
+func loggerWithUserID(logger log.Logger, userID string) log.Logger {
+	if userID == "" {
+		return logger
```

Comment:
should we `level.Warn` log if user ID was empty?

Reply:
We can add it but not sure if it will help much since it is expected to have an empty user id in the common index set.

Reply:
` it is expected to have an empty user id in the common index set.` can you elaborate? I don't understand still

Reply:
Sorry for the confusion, my point is just that it is okay to have an empty user id since it is used for managing the common index set which belongs to all the users and not a specific user.